fedimint-client-rpc/src/lib.rs

```diff
@@ -0,0 +1,311 @@
+use std::collections::HashMap;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time::Duration;
+
+use async_stream::try_stream;
+use fedimint_client::module::IClientModule;
+use fedimint_client::secret::{PlainRootSecretStrategy, RootSecretStrategy};
+use fedimint_client::ClientHandleArc;
+use fedimint_core::db::Database;
+use fedimint_core::invite_code::InviteCode;
+use fedimint_core::task::{MaybeSend, MaybeSync};
+use fedimint_core::util::{BoxFuture, BoxStream};
+use fedimint_core::{apply, async_trait_maybe_send};
+use fedimint_ln_client::{LightningClientInit, LightningClientModule};
+use fedimint_mint_client::{MintClientInit, MintClientModule};
+use futures::future::{AbortHandle, Abortable};
+use futures::StreamExt;
+use serde::{Deserialize, Serialize};
+use tokio::sync::Mutex;
+use tracing::info;
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct RpcRequest {
+    pub request_id: u64,
+    #[serde(flatten)]
+    pub kind: RpcRequestKind,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum RpcRequestKind {
+    JoinFederation {
+        invite_code: String,
+        client_name: String,
+    },
+    OpenClient {
+        client_name: String,
+    },
+    CloseClient {
+        client_name: String,
+    },
+    ClientRpc {
+        client_name: String,
+        module: String,
+        method: String,
+        payload: serde_json::Value,
+    },
+    CancelRpc {
+        cancel_request_id: u64,
+    },
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct RpcResponse {
+    pub request_id: u64,
+    #[serde(flatten)]
+    pub kind: RpcResponseKind,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum RpcResponseKind {
+    Data { data: serde_json::Value },
+    Error { error: String },
+    Aborted {},
+    End {},
+}
+
+pub trait RpcResponseHandler: MaybeSend + MaybeSync {
+    fn handle_response(&self, response: RpcResponse);
+}
+
+#[apply(async_trait_maybe_send!)]
+pub trait DatabaseFactory: MaybeSend + MaybeSync + 'static {
+    async fn create_database(&self, name: &str) -> anyhow::Result<Database>;
```

Comment:
Is this "open"? I mean - it is expected to open previous database if one was used before under same name?

Reply:
oh right this is open or create